pkgs/development/mobile/androidenv/mkrepo.sh

```diff
@@ -6,14 +6,16 @@ set -e
 pushd "$(dirname "$0")" &>/dev/null || exit 1
 
 echo "Writing repo.json" >&2
-ruby mkrepo.rb \
+REPO_JSON="$(cat ./repo.json)"
+echo $REPO_JSON | ruby mkrepo.rb \
```

Comment:
```suggestion
cat ./repo.json | ruby mkrepo.rb \
```

Reply:
In fact, I considered `cat` in the first place, and it took me a while to notice `ruby` cannot handle the STDIN from `cat`! The worst thing is that it doesn't raise an exception/error, the STDIN is just empty! My mental model is that `ruby` is slower than `cat`, so there's a back pressure, which non of them handled it, but when you load the entire data into RAM, there's no pressure and ruby can handle it. Let me know if there's any other way around.

Reply:
You could just load the file from within the ruby script.

Reply:
I have two reasons to not doing it:
 - I'm not the creator of `mkrepo.sh`, or `mkrepo.rb` files, but as far as I understand from their implementation, there's an abstraction is going on. The `mkrepo.rb` is supposed to be agnostic about the where the input is coming from and where the output is going to.
 - My second reason is that I already tried that, and I tried it again after your suggestion, but without success! This is my code
 ```ruby
  input_json = ""
  if File.file? 'repo.json'
    $stderr.write('yes')
  end
  File.foreach('repo.json') { |line|
      input_json += line
  }
  $stderr.write(input_json)
 ```
Its output is just
```console
yes%
```
so `input_json` is empty! Even that `File.foreach` is supposed to be okay with very long files. I have no clue why it's not working! Ruby is not the language that I work with everyday, so maybe I missed something! @numinit I noticed you are much more experienced than me with Ruby. Do you have any idea?

Reply:
After I noticed what's `sponge`, that @SuperSandro2000 suggested below, I could fix it in https://github.com/NixOS/nixpkgs/pull/227666/commits/de2679819e0ab01b9488b151efe4a2d433f1fa25. So in the end, it was not the Ruby problem, but bash is to blame! Thanks @SuperSandro2000.